File: packages/amplify_datastore/android/src/main/kotlin/com/amazonaws/amplify/amplify_datastore/types/model/FlutterSubscriptionEvent.kt
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

package com.amazonaws.amplify.amplify_datastore.types.model

import com.amplifyframework.core.model.SerializedModel

data class FlutterSubscriptionEvent(val serializedModel: SerializedModel, val eventType: String) {
    private val serializedItem: Map<String, Any> =
        FlutterSerializedModel(serializedModel = serializedModel).toMap()

    fun toMap(): Map<String, Any> {
        return mapOf(
            "item" to serializedItem,
            "eventType" to eventType
        )
    }
}

File: packages/amplify_datastore/ios/Classes/types/model/FlutterSubscriptionEvent.swift
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import Flutter
import Foundation
import Amplify

struct FlutterSubscriptionEvent {
    let item: FlutterSerializedModel
    let eventType: EventType
    public init(item: FlutterSerializedModel, eventType: EventType) {
        self.item = item
        self.eventType = eventType
    }

    public func toJSON(modelSchemaRegistry: FlutterSchemaRegistry, customTypeSchemaRegistry: FlutterSchemaRegistry, modelName: String) throws -> [String: Any] {
        return [
            "item": try item.toMap(
                modelSchemaRegistry: modelSchemaRegistry,
                customTypeSchemaRegistry: customTypeSchemaRegistry,
                modelName: modelName
            ),
            "eventType": eventType.rawValue
        ]
    }
}

enum EventType: String {
    case create
    case update
    case delete
}
